grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/GrpcClientFactory.java

```diff
@@ -95,13 +98,14 @@ public Object newClient(ClientBuilderParams params) {
 
         final HttpClient httpClient = newHttpClient(params);
 
-        // TODO(ikhoon): Support gjson with Kotlin coroutine stub once
-        //               https://github.com/grpc/grpc-kotlin/pull/63 is released
-        final MessageMarshaller jsonMarshaller =
-                GrpcSerializationFormats.isJson(serializationFormat) ?
-                GrpcJsonUtil.jsonMarshaller(
-                        stubMethods(clientType),
-                        options.get(GrpcClientOptions.JSON_MARSHALLER_CUSTOMIZER)) : null;
+        GrpcJsonMarshaller jsonMarshaller = null;
+        if (GrpcSerializationFormats.isJson(serializationFormat)) {
+            jsonMarshaller = options.get(GrpcClientOptions.GRPC_JSON_MARSHALLER);
+            if (jsonMarshaller == NoopJsonMarshaller.get()) {
+                jsonMarshaller = new DefaultJsonMarshaller(GrpcJsonUtil.jsonMarshaller(
+                        stubMethods(clientType), options.get(GrpcClientOptions.JSON_MARSHALLER_CUSTOMIZER)));
```

Comment:
This makes me think `GRPC_JSON_MARSHALLER` and `JSON_MARSHALLER_CUSTOMIZER` are not orthogonal. Should we merge these two options into one, e.g. `GRPC_JSON_MARSHALLER_FACTORY` (`Function<Class<?>, GrpcJsonMarshaller>`)? The default value of the new option could be something like:

```java
clientType -> GrpcJsonMarshaller.of(clientType)
```

where `GrpcJsonMarshaller.of()` calls something like `new DefaultJsonMarshaller(GrpcJsonUtil.jsonMarshaller(stubMethods(clientType))`.

If a user wants to customize the marshaller, the user could specify an alternative factory like this:

```java
clientType -> GrpcJsonMarshaller.builder(clientType)
                                ...
                                .build()
```

For ScalaPB and others, it could be `_ -> ScalaPBJsonMarshaller()`.

What do you think, @ikhoon and @anuraaga ?

Reply:
That was my most considering part - JSON marshaller and JSON marshaller customizer. I want to merge the extension points into one.
`GRPC_JSON_MARSHALLER_FACTORY` looks good to me. 👍